celery/utils/time.py

```diff
@@ -300,7 +300,10 @@ def make_aware(dt, tz):
 
 def localize(dt, tz):
     """Convert aware :class:`~datetime.datetime` to another timezone."""
-    dt = dt.astimezone(tz)
+    if is_naive(dt):  # Ensure timezone aware datetime
```

Comment:
I think this code section is what `maybe_make_aware` performs, isn't that correct?

Reply:
Not line for line, I am inspecting it right now and I wouldn't feel comfortable changing this call to `maybe_make_aware` that it does the same thing.  This is saying if its is naive, make it aware of the timezone, if its utc call astimezone on the datetime, which is what the old code used to do exclusively.  I originally just deleted the call to astimezone on the datetime, and that also fixes the bug.

Reply:
Ah I see it now, maybe_make_aware call this method of localize which is why its not the same thing .. we don't want localize to end up calling localize.

Reply:
@georgepsarakis Trying to get this cleaned up for possible merge for 4.2.   I agree that it is a similar check to maybe_make_aware but the bug here is that you can't safely call datetime.astimezone on all timezones.   It doesn't make sense to call `maybe_make_aware` here because that just in turn calls `localize` which would create an infinite recursive loop.    Are you suggesting I actually make a different change here?

Reply:
You are right. My comment was mostly on finding out if the logic is duplicated, or a refactoring would be a better approach.